src/Microsoft.AspNetCore.OData/EnableNestedPathsAttribute.cs

```diff
@@ -0,0 +1,150 @@
+                    actionExecutedContext.Result = new NotFoundObjectResult(null);
+                }
+                else if (path.EdmType.TypeKind == EdmTypeKind.Collection || transformedResult.HasCountSegment)
+                {
+                    responseContent.Value = transformedResult.Result;
+                }
+                else
+                {
+                    Type elementType = transformedResult.Result.ElementType;
+
+                    // If we return the IQueryable as the result, then the response will be returned as a collection
+                    // so we have to return a single result. We can either return the materialized single result using SingleOrDefault,
+                    // which will run the query against the underlying data source. Or we can wrap the IQueryable result with
+                    // a SingleResult<T>.
+                    // If the action has [EnableQuery], then we return a SingleResult<T> which allows query options to be
+                    // applied to the IQueryable. If we returned the materialized object, then query options like $expand
+                    // would return null because navigation properties are not fetched from relational databases by default.
+                    // If the action does not have [EnableQuery], then we run the query using SingleOrDefault and return the
+                    // actual object. If we returned a SingleResult<T> in this case, we'd get an error because the serializer
+                    // does not recognize it.
+                    if (actionDescriptor.MethodInfo.GetCustomAttributes<EnableQueryAttribute>().Any())
+                    {
+                        // calls SingleResult.Create<T>(IQueryable<T>)
+                        MethodInfo singleResultCreate = typeof(SingleResult)
+                            .GetMethod("Create", BindingFlags.Public | BindingFlags.Static)
+                            .MakeGenericMethod(transformedResult.Result.ElementType);
+                        responseContent.Value = singleResultCreate.Invoke(null, new[] { transformedResult.Result });
+                    }
+                    else
+                    {
+                        try
+                        {
+                            object singleValue = QueryHelpers.SingleOrDefault(transformedResult.Result, new WebApiActionDescriptor(actionDescriptor));
+
+                            if (singleValue == null)
+                            {
+                                // TODO: is this the best way to return 404
+                                actionExecutedContext.Result = new NotFoundObjectResult(null);
+                            }
+                            else
+                            {
+                                responseContent.Value = singleValue;
+                            }
+                        }
+                        catch (NullReferenceException)
```

Comment:
Which cases would give us this outcome?

Reply:
Would this be ever triggered by collection having nulls?

Reply:
I removed some of the null-checks from the query builder since they cause errors in EF when constructing the query. But this also leaves the chance that nulls might be encountered when materializing the query, for example when accessing the property of a null item, e.g. /Customers/1/Address/City where `Address` is null. In that case a `NullReferenceException` will be thrown, we catch it here and return a 404.